apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource            `json:"source,omitempty"`
+	Reconcile StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
```

Comment:
Yeah, I was considering this too. The tradeoff would be that there's no longer great support in the model for having different event types or for multiple hooks at that point. It does seem like it fits with our approach of optimizing for the simple use-case we know about. It does mean that we'd need to change the model in some way if we wanted to support any other types of operations such as upgrade-specific logic, and that'll be obvious to anyone who looks at it. But, maybe we'd have a completely different data model to support that anyway. I'll think about this a little more

Reply:
I'm finding that the `initContainer` that copies the helm charts in the image to the controller will need to know the path to the charts.

Reply:
> The tradeoff would be that there's no longer great support in the model for having different event types or for multiple hooks at that point.

That's true. It should be fine either way today.

But keep in mind that we'll probably need to change it anyway to accommodate some kind of list of CRDs and/or behaviors. Maybe it could look like following:
```yaml
behaviors:
  source: crossplaneio/generic-controller:latest
  crd:
   - kind: SampleClaim
     apiVersion: example.io/v1alpha1
     path: /samplekust
     engine:
       type: kustomize
       ....
```
or
```yaml
templates:
  source: crossplaneio/generic-controller:latest
  behaviors:
   - kind: SampleClaim
     apiVersion: example.io/v1alpha1
     path: /samplekust
     engine:
       type: kustomize
       ....
```

Reply:
I ended up moving it into the source, so that we start with as little as possible and build up the complexity as we need it.